natlas-agent/natlas/threadscan.py

```diff
@@ -174,18 +175,23 @@ def execute_scan(self, work_item):
             )
 
     def run(self):
-        try:
-            while True:
+        while True:
+            with push_scope() as scope:
+                add_breadcrumb(category='scan_workflow', message='Fetching work', level='info')
                 work_item = self.get_work()
 
                 if not work_item:
                     break
 
-                self.execute_scan(work_item)
-                work_item.complete()
-        except Exception as e:
-            logger.warning(f"Failed to process work item: {e}")
-            capture_exception(e)
+                scope.set_extra("natlas_scan_id", work_item.target_data['scan_id'])
+                add_breadcrumb(category='scan_workflow', message='Starting scan for %s' % work_item.target_data['target'], level='info')
+                try:
+                    self.execute_scan(work_item)
+                    add_breadcrumb(category='scan_workflow', message='Scan completed. Reporting', level='info')
+                    work_item.complete()
+                except BaseException as e:
+                    event_id = capture_exception(e)
+                    logger.warning(f"Failed to process work item: {e}. Sentry event id: {event_id}")
```

Comment:
There's a logic change here. Previously, a throw exception would cause the thread to die, but now it continues on for the next loop. I prefer to let the thread die to let an orchestrator handle it, but we don't have any yet, so I think this change is fine.

Reply:
I do like the behavior of automatically restarting failed threads, but the change to BaseException means even system exiting exceptions can't be thrown. Ideally we should never need to throw system exiting exceptions from within the run, the only times I think the agent needs to system exit is if the version of the api between the server and agent results in an inability to communicate or if the server requires authentication that the agent isn't providing, at which points all future calls are just wasted cycles and wasted bandwidth. 

That said, it looks like I yolo'd it so that if we get an HTTP 403, I just call `os._exit` which is bad but might work to our favor here. Except I guess not because Docker might still automatically try to restart 🤔. I'll want to test it out before I approve it though, just so that we know exactly what the behavior is.

Reply:
I don't see an issue with reporting BaseException. At work, I catch `java.lang.Throwable` and report it to Sentry so I can catch OutOfMemoryError (Java world.) But I do see the report with swallowing all types of exceptions. What should we do here? Re-raise it up so that the thread dies, as it does today or try to smartly handle certain exceptions.

Reply:
I just pulled up the hierarchy for BaseException:
* BaseException
  * Exception
    * ArithmeticError
    * Literally everything
  * GeneratorExit
  * KeyboardInterrupt
  * SystemExit

And after reviewing [this doc](https://help.semmle.com/wiki/pages/viewpage.action?pageId=29394005) it seems that BaseException doesn't provide us with anything better over Exception.

Reply:
Reporting BaseException is fine, maybe something like this:

```
try:
    raise SystemExit("Hello world")
except BaseException as e:
    capture_exception(e)
    if isinstance(e, SystemExit):
        raise SystemExit
    print("gotta go fast")
```

I think instead of SystemExit though we might want to long-term think about a halt that we can do where we stop sending requests to the server, drop an error in the log, and just wait forever. That way the container doesn't die and `--restart=on-failure` or `--restart=always` automatically kicks it off again.